# Spam Detector

In [1]:
#import libraries
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

## Retrieve the Data

The data is located at [https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv](https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [2]:
# Import the data
data = pd.read_csv("https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv")
data.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

In [3]:
# Look at the shape of the data to see number of rows and columns
data.shape

(4601, 58)

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct! 

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

My prediction is that the Random Forest Classifier model will perform better.  Random Forest is better than Logistic Regression at handling complex datasets with a large number of features.  This dataset has a fairly large number of features - it has 57 different features. In many cases, Random Forest Classifier achieves higher accuracy than Logistic Regression due to its ability to capture complex relationships.  

## Split the Data into Training and Testing Sets

In [4]:
# Create the labels set `y` and features DataFrame `X`
y = data["spam"]

X = data.copy()
X = X.drop(columns="spam")

In [5]:
# Check the balance of the labels variable (`y`) by using the `value_counts` function.
y.value_counts()

spam
0    2788
1    1813
Name: count, dtype: int64

In [6]:
# Split the data into X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

## Scale the Features

Use the `StandardScaler` to scale the features data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [7]:
from sklearn.preprocessing import StandardScaler

# Create the StandardScaler instance
scaler = StandardScaler()

In [8]:
# Fit the Standard Scaler with the training data
scaler.fit(X_train)

StandardScaler()

In [9]:
# Scale the training data
X_train_scaled = scaler.transform(X_train)
X_train_scaled

array([[-0.35811925, -0.16744248, -0.56071981, ..., -0.13974678,
        -0.23023328, -0.37548424],
       [-0.35811925, -0.16744248, -0.56071981, ..., -0.13640572,
        -0.23023328, -0.46466704],
       [ 0.21317665, -0.16744248, -0.22391235, ..., -0.01186062,
         0.13008597,  1.25168127],
       ...,
       [-0.35811925, -0.16744248,  0.94500764, ..., -0.10138496,
        -0.16385868, -0.06418577],
       [-0.35811925, -0.16744248,  1.16294188, ..., -0.1351981 ,
        -0.20178702, -0.01707033],
       [-0.35811925, -0.16744248, -0.56071981, ..., -0.15524447,
        -0.23971537, -0.42933046]])

In [10]:
# Transform the test dataset based on the fit from the training set
X_test_scaled = scaler.transform(X_test)
X_test_scaled

array([[-0.35811925, -0.16744248, -0.56071981, ...,  0.07078032,
        -0.09748408, -0.12139738],
       [-0.35811925, -0.16744248, -0.56071981, ..., -0.11736569,
        -0.14963555, -0.32836879],
       [-0.35811925,  0.24309223, -0.56071981, ..., -0.01628853,
        -0.01214531, -0.27620527],
       ...,
       [ 1.01971205, -0.16744248,  0.25158053, ..., -0.12465162,
        -0.1875639 , -0.41418621],
       [-0.35811925, -0.16744248, -0.56071981, ..., -0.15939061,
        -0.24919745, -0.47476321],
       [-0.35811925,  0.1966166 , -0.56071981, ..., -0.11410514,
        -0.20178702, -0.31154185]])

## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [11]:
# Create a Logistic Regression model
from sklearn.linear_model import LogisticRegression

# Create the Logistic Regression model
lr_model = LogisticRegression(random_state=1)

In [12]:
# Fit the model using X_train_scaled and y_train
lr_model.fit(X_train_scaled, y_train)

LogisticRegression(random_state=1)

In [13]:
# Make and save testing predictions with the saved logistic regression model using the test data
lr_testing_predictions = lr_model.predict(X_test_scaled)

# Review the predictions
lr_testing_predictions

array([0, 0, 1, ..., 0, 0, 1], dtype=int64)

In [14]:
# Calculate the accuracy score by evaluating `y_test` vs. `lr_testing_predictions`.
accuracy_score(y_test, lr_testing_predictions)

0.9278887923544744

## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [15]:
# Create a Random Forest Classifier model
from sklearn.ensemble import RandomForestClassifier

# Create the Random Forest Classifier model
rf_model = RandomForestClassifier(random_state=1)


In [16]:
# Fit the model using X_train_scaled and y_train
rf_model.fit(X_train_scaled, y_train)

RandomForestClassifier(random_state=1)

In [17]:
# Make and save testing predictions with the saved logistic regression model using the test data
rf_testing_predictions = rf_model.predict(X_test_scaled)

# Review the predictions
rf_testing_predictions

array([1, 1, 1, ..., 1, 0, 1], dtype=int64)

In [18]:
# Calculate the accuracy score by evaluating `y_test` vs. `rf_testing_predictions`.
accuracy_score(y_test, rf_testing_predictions)

0.9669852302345786

## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.

The Random Forest Classifier model performs better than the Logistic Regression model. I had predicted that the Random Model Classifier model would in fact perform better. The accuracy score with the Random Forest Classifier is approx. 96.7% compared to Logistic Regression with an accuracy score of approx. 92.8%.  This makes the performance of the Random Forest Classifier considerably better.  After experimenting with the two models, you would definitely choose the Random Forest Classifier as the better model.  

In order to be sure that the Random Forest Classifier is the best model in this case, you may want to experiment with a selection of other models such as the Extremely Randomized Trees model, which is another ensemble learning model that sometimes performs better depending on the dataset.